### PR TITLE
fix fields and-or logic

### PR DESCRIPTION
**Pull request by ns3777k**
issue https://github.com/sequelize/sequelize/issues/3767

### CODE REVIEW

File: lib/dialects/abstract/query-generator.js [hunk 1]
```diff
@@ -1789,18 +1789,25 @@ var QueryGenerator = {
...
 
-        value = (value.$or || value.$and).map(function (_value) {
-          return self.whereItemQuery(key, _value, options);
-        }).filter(function (item) {
-          return item && item.length;
+      if (_.isPlainObject(value)) {
+        var where = [];
+        _.forOwn(value, function (_value, key) {
+          where.push(_.zipObject([key], [_value]));
         });
```
Comment (janmeier) [inline lib/dialects/abstract/query-generator.js]: Pretty clever use of zip here to assign a variable as key without creating the object first. I like it :-)

Comment (ns3777k) [inline lib/dialects/abstract/query-generator.js]: thx :-)

Comment (mickhansen) [inline lib/dialects/abstract/query-generator.js]: Using a reduce would kill the need for an intermediate `where` variable.

Comment (mickhansen) [inline lib/dialects/abstract/query-generator.js]: Or could really just use `_.mapValues` here.

Comment (ns3777k) [inline lib/dialects/abstract/query-generator.js]: changed to `_.reduce`. works.

`_.mapValues` ? If i'm not missing anything it only changes the values keeping same keys.
but we need to transform the keys too. i.e. incoming object from:
```
{ '$lt': 100, '$eq': null }
```
to:
```
[ { '$lt': 100 }, { '$eq': null } ]
```

**Comment by ns3777k** [comment]
added test for $and

File: lib/dialects/abstract/query-generator.js [hunk 2]
```diff
@@ -1789,18 +1789,24 @@ var QueryGenerator = {
...
 
-        value = (value.$or || value.$and).map(function (_value) {
-          return self.whereItemQuery(key, _value, options);
-        }).filter(function (item) {
-          return item && item.length;
-        });
-
-        return value.length ? '('+value.join(binding)+')' : undefined;
+      if (_.isPlainObject(value)) {
+        value = _.reduce(value, function (result, _value, key) {
```
Comment (mickhansen) [inline lib/dialects/abstract/query-generator.js]: Beautiful :)

File: test/unit/sql/where.test.js [hunk 3]
```diff
@@ -259,6 +268,15 @@ suite(Support.getTestDialectTeaser('SQL'), function() {
           default: "([name] LIKE '%hello' AND [name] LIKE 'hello%')"
         });
 
+        testsql('rank', {
+          $and: {
```
Comment (mickhansen) [inline test/unit/sql/where.test.js]: This should work simply without the `$and`, but a test is always good :)

**Comment by mickhansen** [comment]
Great work

**Pull request merged**